### ISSUE TITLE
Support "nix-collect-garbage -d --keep-generations 2"

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nix#175]
Would be very useful to be able to keep last X generation when doing garbage collection.

**Comment by musteresel** [comment]
Has there ever been any work to implement this feature? What parts of nix would one need to touch?

**Comment by mogorman** [comment]
i actually wrote a pr for that #767

**Comment by jackjennings** [comment]
This was added in the previously linked PR: https://github.com/NixOS/nix/pull/767

**Comment by bbigras** [comment]
Does `nix-collect-garbage` has something like `--keep-generations`?

It seems that the PR added `--keep-generations` to `nix-env`. I have no idea how to use it with `nix-collect-garbage` to clean system stuff.

**Comment by alarsyo** [comment]
Could this be reopened as it would be really useful in `nix-collect-garbage`? Currently the only way to delete the `k` last NixOS generations is to run the following command:

```sh
nix-env --delete-generations +k --profile /nix/var/nix/profiles/system
```

making this usage incompatible with all the `nix.gc` NixOS module options, which only allow passing flags to `nix-collect-garbage`.

**Comment by OleMussmann** [comment]
It could be dangerous for systems that have longer down-times to have their generations scrubbed by time. A buffer of a certain number of generations could prevent that. Please consider re-opening this issue.

**Comment by EternityForest** [comment]
If this is already solved, could we get some documentation on how to use it?

**Comment by miku86** [comment]
Any news on this?

I have a system that gets updated only once every some months.
Using `--delete-older-than` is not very practical here, because most config changes are at the same day, so I can either delete all of them or none.

**Comment by Pandapip1** [comment]
I'd recommend just opening a new issue. It seems unlikely that a maintainer will just stumble across this issue and reopen it.

**Comment by Ericson2314** [comment]
#9894 should help, and I am sorry to say I am just behind on reviewing that because of other tasks.

**Comment by Pandapip1** [comment]
> It seems unlikely that a maintainer will just stumble across this issue and reopen it.

I stand corrected.

**Comment by eliasjurkeit** [comment]
@Ericson2314 dupe of #9455.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
